packages/gitgraph-js/src/gitgraph.ts

```diff
@@ -446,164 +544,128 @@ function createGitgraph(
-
-    const tooltip = commit.renderTooltip
-      ? commit.renderTooltip(commit)
-      : createTooltip(commit);
-
-    $tooltip = createG({
-      translate: getWithCommitOffset(commit),
-      children: [tooltip],
-    });
+    function setChildrenForeignObjectHeight(node: Node): void {
+      if (node.nodeName === "foreignObject") {
+        // We have to access the first child's parentElement to retrieve
+        // the Element instead of the Node => we can compute dimensions.
+        const foreignObject = node.firstChild && node.firstChild.parentElement;
+        if (!foreignObject) return;
 
-    svg.firstChild.appendChild($tooltip);
-  }
+        // Force the height of the foreignObject (browser issue)
+        foreignObject.setAttribute(
+            "height",
+            getMessageHeight(foreignObject.firstElementChild).toString(),
+        );
+      }
 
-  function getWithCommitOffset({ x, y }: Coordinate): Coordinate {
-    return { x, y: commitYWithOffsets[y] || y };
+      node.childNodes.forEach(setChildrenForeignObjectHeight);
+    }
   }
 
-  function setBranchLabelRef(commit: Commit, branchLabels: SVGGElement): void {
-    if (!commitsElements[commit.hashAbbrev]) {
-      initCommitElements(commit);
+  protected setBranchLabelRef(commit: Commit, branchLabels: SVGGElement): void {
+    if (!this.commitsElements[commit.hashAbbrev]) {
+      this.initCommitElements(commit);
     }
 
-    commitsElements[commit.hashAbbrev].branchLabel = branchLabels;
+    this.commitsElements[commit.hashAbbrev].branchLabel = branchLabels;
   }
 
-  function setMessageRef(commit: Commit, message: SVGGElement | null): void {
-    if (!commitsElements[commit.hashAbbrev]) {
-      initCommitElements(commit);
+  protected setMessageRef(commit: Commit, message: SVGGElement | null): void {
+    if (!this.commitsElements[commit.hashAbbrev]) {
+      this.initCommitElements(commit);
     }
 
-    commitsElements[commit.hashAbbrev].message = message;
+    this.commitsElements[commit.hashAbbrev].message = message;
   }
 
-  function setTagRef(commit: Commit, tag: SVGGElement): void {
-    if (!commitsElements[commit.hashAbbrev]) {
-      initCommitElements(commit);
+  protected setTagRef(commit: Commit, tag: SVGGElement): void {
+    if (!this.commitsElements[commit.hashAbbrev]) {
+      this.initCommitElements(commit);
     }
 
-    commitsElements[commit.hashAbbrev].tags.push(tag);
+    this.commitsElements[commit.hashAbbrev].tags.push(tag);
   }
 
-  function initCommitElements(commit: Commit): void {
-    commitsElements[commit.hashAbbrev] = {
+  protected initCommitElements(commit: Commit): void {
+    this.commitsElements[commit.hashAbbrev] = {
       branchLabel: null,
       tags: [],
       message: null,
     };
   }
 }
 
+function createGitgraph(
+  graphContainer: HTMLElement,
+  options?: GitgraphOptions,
+  renderer?: Renderer,
+) {
+
+  const gitgraph = new GitgraphCore(options);
+  if (renderer) {
+    renderer.gitgraph = gitgraph;
```

Comment:
I'm surprised with this line. We should set attributes to an instance on the fly. 

There are different ways to solve this. I think the best way to do so is to delay the instantiation of the `renderer`, since we need to pass it the `gitgraph` instance anyway. So maybe instead provide something like that:

```ts
function createGitgraph(
  graphContainer: HTMLElement,
  options?: GitgraphOptions,
  createRenderer?: (graphContainer: HTMLElement, gitgraph: GitgraphCore) => Renderer,
) {
  const gitgraph = new GitgraphCore(options);
  const renderer = createRenderer(graphContainer, gitgraph);

  // React on gitgraph updates to re-render the graph.
  gitgraph.subscribe((data) => {
    renderer.render(data, true);
  });

  // Return usable API for end-user.
  return gitgraph.getUserApi();
}
```

Or, if we want to enforce classes:

```ts
function createGitgraph(
  graphContainer: HTMLElement,
  options?: GitgraphOptions,
  GraphRenderer?: typeof Renderer,
) {
  const gitgraph = new GitgraphCore(options);
  const renderer = new GraphRenderer(graphContainer, gitgraph);

  // React on gitgraph updates to re-render the graph.
  gitgraph.subscribe((data) => {
    renderer.render(data, true);
  });

  // Return usable API for end-user.
  return gitgraph.getUserApi();
}
```

Reply:
Yes! good catch, this line bothered me also.  Of the two solutions you propose, I like the `createRenderer` function better because the caller may want to add some more arguments to their custom renderer class constructor.

Also,  as the code is now, a naive caller may want to re-use the same renderer instance to multiple git-graph instances which will, of course, fail horribly.  It's much better to enforce creating a new renderer instance inside the `createGitgraph` function even for this reason alone.